18.md

```diff
@@ -1,56 +1,96 @@
 LUD-18: Payer identity in `payRequest` protocol.
 ================================================
 
-`author: akumaigorodski`
+`author: akumaigorodski` `author: hampus_s` `author: fiatjaf`
 
 ---
 
-## Specifying payer identity before sending a payment
+The idea here is that the payer can identify itself when paying. The scope varies from free-form names (or [LUD-16](16.md) internet identifiers) useful for loose identification in comments or similar things, to cryptographic keys for proof of payment, to authentication keys for future login into a `SERVICE`'s website after a payment and other use cases.
 
-Payer MAY attach a `payerid` query parameter to LNURL-PAY callback with value set to JSON array:
+Aside from that, the payer ids are also committed to the invoice by means of the `descriptionHash` property, which ensures strong cryptographic guarantees for proof of payment.
+
+## 1. `payerIDs` record
+
+If `SERVICE` wants to get one or more types of payer identities from `WALLET` then it must alter its JSON response to the first callback to include a `payerIDs` field, as follows:
+
+```diff
+ {
+   callback: String,
+   maxSendable: MilliSatoshi,
+   minSendable: MilliSatoshi,
+   metadata: String,
++  payerIDs: {
++    "name": { "mandatory": Boolean },
```

Comment:
Are all of these really necessary to be in the payload? Seems excessive. `payerIDs` should be a freeform  specification requirement defined by `SERVICE`. with a schema:`{ [fieldName:string]: { mandatory?:boolean, format?:string}}`

Reply:
Everything isn't necessary in the payload. Note that we're not using Typescript definitions in LNURL.
This is just listing every possible payerid.

Reply:
If `SERVICE` requires the payer's `name` and would also like their `email` (optional), the `payerIDs` field would be:

```
payerIDs: {
  "name": { mandatory: true },  <--- if not included by wallet, service will fail the payment
  "email": { mandatory: false } <--- just a nice to have
}
```

Other fields like `auth`  that `SERVICE` does not care about are simply not included.